File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-option-shared-5-4/CalendarSpread.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class CalendarSpread : ISerialized
    {
        public CalendarSpread(XmlNode xmlNode)
        {
            XmlNode expirationDateTwoNode = xmlNode.SelectSingleNode("expirationDateTwo");
            
            if (expirationDateTwoNode != null)
            {
                if (expirationDateTwoNode.Attributes["href"] != null || expirationDateTwoNode.Attributes["id"] != null) 
                {
                    if (expirationDateTwoNode.Attributes["id"] != null) 
                    {
                        expirationDateTwoIDRef_ = expirationDateTwoNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(expirationDateTwoNode);
                        IDManager.SetID(expirationDateTwoIDRef_, ob);
                    }
                    else if (expirationDateTwoNode.Attributes["href"] != null)
                    {
                        expirationDateTwoIDRef_ = expirationDateTwoNode.Attributes["href"].Value;
                    }
                    else
                    {
                        expirationDateTwo_ = new AdjustableOrRelativeDate(expirationDateTwoNode);
                    }
                }
                else
                {
                    expirationDateTwo_ = new AdjustableOrRelativeDate(expirationDateTwoNode);
                }
            }
            
        
        }
        
    
        #region ExpirationDateTwo_
        private AdjustableOrRelativeDate expirationDateTwo_;
        public AdjustableOrRelativeDate ExpirationDateTwo_
        {
            get
            {
                if (this.expirationDateTwo_ != null)
                {
                    return this.expirationDateTwo_; 
                }
                else if (this.expirationDateTwoIDRef_ != null)
                {
                    expirationDateTwo_ = IDManager.getID(expirationDateTwoIDRef_) as AdjustableOrRelativeDate;
                    return this.expirationDateTwo_; 
                }
                else
                {
                      return this.expirationDateTwo_; 
                }
            }
            set
            {
                if (this.expirationDateTwo_ != value)
                {
                    this.expirationDateTwo_ = value;
                }
            }
        }
        #endregion
        
        public string expirationDateTwoIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-option-shared-5-4/CalendarSpread.cpp
// CalendarSpread.cpp 
#include "CalendarSpread.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

CalendarSpread::CalendarSpread(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //expirationDateTwoNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* expirationDateTwoNode = xmlNode->FirstChildElement("expirationDateTwo");

   if(expirationDateTwoNode){expirationDateTwoIsNull_ = false;}
   else{expirationDateTwoIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- expirationDateTwoNode , address : " << expirationDateTwoNode << std::endl;
   #endif
   if(expirationDateTwoNode)
   {
      if(expirationDateTwoNode->Attribute("href") || expirationDateTwoNode->Attribute("id"))
      {
          if(expirationDateTwoNode->Attribute("id"))
          {
             expirationDateTwoIDRef_ = expirationDateTwoNode->Attribute("id");
             expirationDateTwo_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(expirationDateTwoNode));
             expirationDateTwo_->setID(expirationDateTwoIDRef_);
             IDManager::instance().setID(expirationDateTwoIDRef_,expirationDateTwo_);
          }
          else if(expirationDateTwoNode->Attribute("href")) { expirationDateTwoIDRef_ = expirationDateTwoNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { expirationDateTwo_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(expirationDateTwoNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<AdjustableOrRelativeDate> CalendarSpread::getExpirationDateTwo()
{
   if(!this->expirationDateTwoIsNull_){
        if(expirationDateTwoIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(expirationDateTwoIDRef_));
        }else{
             return this->expirationDateTwo_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
}

